src/bc_protocol/adpcm.rs

```diff
@@ -0,0 +1,226 @@
+
+                let raw_sample;
+                /* == Non approxiate version ===
+                // This is the full maths version
+                // We don't use this one as we need to match the way the encoder
+                // works if we want to use the state stored in the header.
+                // I have Left it here as it is easier to understand then the bit shift version below
+                let inibble = nibble.signed();
+
+                // Calculate the delta (which is really what adpcm is all about)
+                // Adaptive **Differential** PCM
+                // Differential: Becuase its all about the difference (gradient)
+                let diff = (step as isize) * (inibble) / 2 + (step as isize) / 8;
+
+                // Eulers approxiation
+                // Sample = Previous_Sample + difference*step_size
+                raw_sample = last_output + diff;
+                */
+
+                // === Approximate version ==
+                // Approximate form uses bit shift operators.
+                // This is a legacy of the days when mult/divides were expensive
+                // It is also the format used on low end CPUs like cameras
+                let mut diff = step >> 3;
+                if (unibble & 0b0100) == 0b0100 {
+                    diff += step;
+                }
+                if (unibble & 0b0010) == 0b0010 {
+                    diff += step >> 1;
+                }
+                if (unibble & 0b0001) == 0b0001 {
+                    diff += step >> 2;
+                }
+                // Sign test
+                if (unibble & 0b1000) == 0b1000 {
+                    raw_sample = last_output - (diff as isize);
+                } else {
+                    raw_sample = last_output + (diff as isize);
+                }
+
+                // Specifications say: Clamp it in max sample range -context.max_sample_size..context.max_sample_size
+                let sample = match raw_sample {
+                    value if value > context.max_sample_size - 1 => context.max_sample_size - 1,
+                    value if value < -context.max_sample_size => -context.max_sample_size,
+                    value => value,
+                };
+
+                // PCM is really in i16 range
+                // Some formats e.g. OKI are not in the full PCM range of values
+                // To convert we must scale it to the i16 range
+                // We also cast to i16 at this point ready for the conversion to u8 bytes of the output
+                let scaled_sample = (sample as isize * (i16::MAX as isize)
+                    / (context.max_sample_size - 1) as isize)
+                    as i16;
+
+                // Get the results in bytes
+                result.extend(scaled_sample.to_le_bytes().iter());
+
+                // Increment the step index
+                step_index = step_index as isize + context.changes[unibble];
```

Comment:
This slice is okay, because you've constrained the index to be 0..16 by dint of the nibble bitmask.

Reply:
Yes this part here should all be in range be nature of the constraints